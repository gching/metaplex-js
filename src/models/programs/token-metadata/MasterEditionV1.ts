import { PublicKey } from '@solana/web3.js';
import BN from 'bn.js';
import { deserializeUnchecked } from 'borsh';
import { BorshSchema } from '../../utils/types';
import { MetadataKey } from './MetadataKey';
import { AccountWithPublicKey } from '../../../utils/privateTypes';

export class MasterEditionV1 {
  static SCHEMA: BorshSchema = new Map([
    [
      MasterEditionV1,
      {
        kind: 'struct',
        fields: [
          ['key', 'u8'],
          ['supply', 'u64'],
          ['maxSupply', { kind: 'option', type: 'u64' }],
          ['printingMint', 'pubkey'],
          ['oneTimePrintingAuthorizationMint', 'pubkey'],
        ],
      },
    ],
  ]);

  key = MetadataKey.MasterEditionV1;
  accountPublicKey?: PublicKey;
  supply: BN;
  maxSupply?: BN;
  /// Can be used to mint tokens that give one-time permission to mint a single limited edition.
  printingMint: PublicKey;
  /// If you don't know how many printing tokens you are going to need, but you do know
  /// you are going to need some amount in the future, you can use a token from this mint.
  /// Coming back to token metadata with one of these tokens allows you to mint (one time)
  /// any number of printing tokens you want. This is used for instance by Auction Manager
  /// with participation NFTs, where we dont know how many people will bid and need participation
  /// printing tokens to redeem, so we give it ONE of these tokens to use after the auction is over,
  /// because when the auction begins we just dont know how many printing tokens we will need,
  /// but at the end we will. At the end it then burns this token with token-metadata to
  /// get the printing tokens it needs to give to bidders. Each bidder then redeems a printing token
  /// to get their limited editions.
  oneTimePrintingAuthorizationMint: PublicKey;

  constructor(args: Omit<MasterEditionV1, 'key'>) {
    this.accountPublicKey = args.accountPublicKey;
    this.supply = args.supply;
    this.maxSupply = args.maxSupply;
    this.printingMint = args.printingMint;
    this.oneTimePrintingAuthorizationMint =
      args.oneTimePrintingAuthorizationMint;
  }

  static decode(accountWithPublicKey: AccountWithPublicKey): MasterEditionV1 {
    const masterEdition = deserializeUnchecked(
      MasterEditionV1.SCHEMA,
      MasterEditionV1,
      accountWithPublicKey.account.data,
    ) as MasterEditionV1;

    return new MasterEditionV1({
      ...masterEdition,
      accountPublicKey: accountWithPublicKey.pubkey,
    });
  }
}
