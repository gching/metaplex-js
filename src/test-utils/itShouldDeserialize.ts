import { PublicKey } from '@solana/web3.js';
import { serialize } from 'borsh';
import { BorshSchema } from '../models/utils/types';

export function itShouldDeserialize<E>(
  cls: {
    SCHEMA: BorshSchema;
    // eslint-disable-next-line @typescript-eslint/ban-types
    decode: Function;
  },
  received: E,
  accountPubKey = PublicKey.default,
): void {
  const data = Buffer.from(serialize(cls.SCHEMA, received));

  const deserialized = cls.decode({
    pubkey: accountPubKey,
    account: {
      data,
    },
  }) as E;

  expect(deserialized).toDeepEqual(received);
}
