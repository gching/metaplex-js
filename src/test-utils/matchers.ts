import { isEqualWith } from 'lodash';
import BN from 'bn.js';

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    interface Matchers<R> {
      toDeepEqual: <E = any>(expected: E) => R;
    }
  }
}

expect.extend({
  toDeepEqual<E = any>(received: E, expected: E) {
    const pass = isEqualWith(received, expected, (objValue, otherValue) => {
      if (BN.isBN(objValue) && BN.isBN(otherValue)) {
        return objValue.eq(otherValue);
      }
      return undefined;
    });

    if (pass) {
      return {
        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
        message: () => `expected ${received} not to deep equal ${expected}`,
        pass: true,
      };
    }

    return {
      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
      message: () => `expected ${received} to deep equal ${expected}`,
      pass: false,
    };
  },
});

export default undefined;
