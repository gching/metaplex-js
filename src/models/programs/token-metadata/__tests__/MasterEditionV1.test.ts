import { PublicKey } from '@solana/web3.js';
import BN from 'bn.js';
import { itShouldDeserialize } from '../../../../test-utils/itShouldDeserialize';
import { MasterEditionV1 } from '../MasterEditionV1';

describe('MasterEditionV1', () => {
  it('should instantiate with the correct values', () => {
    const inst = new MasterEditionV1({
      accountPublicKey: PublicKey.default,
      oneTimePrintingAuthorizationMint: PublicKey.default,
      supply: new BN(0),
      printingMint: PublicKey.default,
    });

    expect(inst).toMatchSnapshot();
  });

  it('should deserialize with the correct values', () => {
    const inst = new MasterEditionV1({
      accountPublicKey: PublicKey.default,
      oneTimePrintingAuthorizationMint: PublicKey.default,
      supply: new BN(0),
      printingMint: PublicKey.default,
    });

    itShouldDeserialize(MasterEditionV1, inst);
  });
});
