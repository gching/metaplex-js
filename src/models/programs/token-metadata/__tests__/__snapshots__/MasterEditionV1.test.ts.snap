// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MasterEditionV1 should instantiate with the correct values 1`] = `
MasterEditionV1 {
  "accountPublicKey": PublicKey {
    "_bn": "00",
  },
  "key": 2,
  "maxSupply": undefined,
  "oneTimePrintingAuthorizationMint": PublicKey {
    "_bn": "00",
  },
  "printingMint": PublicKey {
    "_bn": "00",
  },
  "supply": "00",
}
`;
